src/hooks/with-auth.tsx

```diff
@@ -0,0 +1,67 @@
+import React, { Component } from 'react';
+import { NextApiRequest, NextApiResponse, NextPage, NextPageContext } from 'next';
+import Router from 'next/router';
+
+import { NullableUserProfile, useUser } from './use-user';
+import { ISignInWithAuth0 } from '../instance';
+import { createLoginUrl } from '../utils/url-helpers';
+
+type RedirectToLoginProps = {
+  render: () => React.ReactElement;
+};
+
+class RedirectToLogin extends Component<RedirectToLoginProps> {
+  public constructor(props: RedirectToLoginProps) {
+    super(props);
+  }
+
+  public componentDidMount(): void {
+    window.location.assign(createLoginUrl(Router.pathname));
+  }
+
+  public render(): React.ReactElement {
+    return this.props.render();
+  }
+}
+
+type AuthenticatedProps = React.PropsWithChildren<{ user: NullableUserProfile }>;
+
+export default function withAuth(
+  InnerComponent: React.ElementType,
+  redirect: () => React.ReactElement,
+  instance: ISignInWithAuth0
+): NextPage<AuthenticatedProps> {
+  const Authenticated: NextPage<AuthenticatedProps> = (props) => {
+    const { user } = useUser();
+
+    if (!user) {
+      return <RedirectToLogin render={redirect} />;
+    }
+
+    return <InnerComponent {...props} user={user} />;
+  };
+
+  Authenticated.getInitialProps = async (context: NextPageContext): Promise<AuthenticatedProps> => {
```

Comment:
Also, since we're targeting 9.4+ we should probably use `getServerSideProps`

> If you're using Next.js 9.3 or newer, we recommend that you use getStaticProps or getServerSideProps instead of getInitialProps.
https://nextjs.org/docs/api-reference/data-fetching/getInitialProps